src/snowflake/snowpark/dataframe_transform_functions.py

```diff
@@ -120,3 +121,102 @@ def moving_agg(
 
         return agg_df
+
+    def encode(
+        self,
+        cols: List[str],
+        encoding_types: List[str],
+        targets: Optional[List[str]] = None,
+        col_formatter: Callable[[str, str], str] = _default_col_formatter,
+    ) -> "snowflake.snowpark.dataframe.DataFrame":
+        """
+        Encodes columns in the DataFrame using specified encoding types.
+
+        Args:
+            cols: A list of column names to encode.
+            encoding_types: A list of encoding types corresponding to each column. Supported types include 'label', 'one_hot', and 'target'.
+            targets: An optional list of target columns for 'target' encoding. For non-target encodings, None should be specified.
+            col_formatter: An optional function for formatting output column names. The default format is '<input_col>_<encoding_type>'.
+                        This function takes two arguments: 'input_col' (str) for the column name and 'encoding_type' (str) for the type of encoding,
+                        and returns a formatted string for the column name.
+
+        Returns:
+            A Snowflake DataFrame with additional columns corresponding to each encoded column.
+
+        Raises:
+            ValueError: If an unsupported encoding type is specified or if required parameters for specific encoding types are not provided.
+            TypeError: If input arguments are not of expected types.
+
+        Example:
+            encoded_df = df.transform.encode(
+                cols=['WEATHER_DESCRIPTION', 'ANOTHER_COLUMN'],
+                encoding_types=['target', 'one_hot'],
+                targets=['DAILY_HIGH_TEMP', None],
+                col_formatter=col_formatter_func
+            )
+        """
+        self._validate_string_list_argument(cols, "cols")
+        self._validate_string_list_argument(encoding_types, "encoding_types")
+        self._validate_formatter_argument(col_formatter)
+
+        if targets is None:
+            targets = [None] * len(cols)
+
+        if not (len(cols) == len(encoding_types)):
+            raise ValueError("Length of cols, encoding_types must be equal.")
+
+        encoded_df = self._df
+        for col_name, encoding_type, target in zip(cols, encoding_types, targets):
+            formatted_col_name = col_formatter(col_name, encoding_type)
+            column = _to_col_if_str(col_name, "transform.encode")
+
+            if encoding_type == "label":
+                # Create a DataFrame with distinct values and their labels
+                distinct_values = (
+                    self._df.select(column)
+                    .distinct()
+                    .with_column(
+                        formatted_col_name, rank().over(Window.order_by(column))
+                    )
+                )
+
+                # Add the label encoding to the df.
+                distinct_values = distinct_values.with_column_renamed(
+                    col_name, "temp_join_key"
```

Comment:
nit: consider name temp col starts with `_` to reduce to chance conflict with user columns.